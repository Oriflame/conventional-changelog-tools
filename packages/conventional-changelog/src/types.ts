import type { Commit, ParserOptions, SemverLevel } from '@oriflame/conventional-changelog-types';

export interface BumpOptions {
  parserOpts: Partial<ParserOptions>;
  whatBump: (commits: Commit[]) => { level: SemverLevel; reason: string };
}
