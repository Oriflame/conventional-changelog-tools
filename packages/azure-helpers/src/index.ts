import { getPersonalAccessTokenHandler, WebApi } from 'azure-devops-node-api';
import { WorkItem } from 'azure-devops-node-api/interfaces/WorkItemTrackingInterfaces';

interface AzureClientOptions {
  serverUrl: string;
  pat: string;
}

export async function createAzureClient({ pat, serverUrl }: AzureClientOptions) {
  const authHandler = getPersonalAccessTokenHandler(pat);
  const webApi = new WebApi(serverUrl, authHandler);
  await webApi.connect();

  return webApi;
}

interface WorkItemUrlOptions {
  workItem: WorkItem;
  organizationUrl: string;
}

export function getWorkItemUrl({
  workItem: { fields = {}, id },
  organizationUrl,
}: WorkItemUrlOptions) {
  return `${organizationUrl}/${fields['System.TeamProject']}/_workitems/edit/${id}`;
}
