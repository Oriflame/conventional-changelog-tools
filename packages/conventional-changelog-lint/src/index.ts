/**
 * @copyright   2020, Oriflame Software
 * @license     https://opensource.org/licenses/MIT
 */

import { GROUPS } from '@oriflame/conventional-changelog-types';

export default {
  rules: {
    'body-max-length': [0],
    'scope-case': [0, 'never', 'start-case'],
    'subject-full-stop': [2, 'always', '.'],
    'subject-case': [2, 'always', 'sentence-case'],
    'type-enum': [2, 'always', GROUPS.flatMap((current) => current.types)],
  },
};
