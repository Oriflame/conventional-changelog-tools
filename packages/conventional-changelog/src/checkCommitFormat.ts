import { CommitType, COMMIT_FORMAT_PREFIX } from '@oriflame/conventional-changelog-types';

export default function checkCommitFormat(
  commit: string,
): { scope: string; type: CommitType } | null {
  const match = new RegExp(`${COMMIT_FORMAT_PREFIX.source}`, 'u').exec(commit);

  if (!match) {
    return null;
  }
  const [, type, scope = ''] = match;

  return {
    scope: scope || '',
    type: type as CommitType,
  };
}
