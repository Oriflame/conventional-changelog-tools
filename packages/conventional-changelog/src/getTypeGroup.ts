import type { CommitType, Group } from '@oriflame/conventional-changelog-types';
import { GROUPS } from '@oriflame/conventional-changelog-types';

export default function getTypeGroup(type: CommitType): Group {
  const group = GROUPS.find((g) => g.types.includes(type));

  if (!group) {
    throw new Error(`Cannot find group for type "${type}".`);
  }

  return group;
}
