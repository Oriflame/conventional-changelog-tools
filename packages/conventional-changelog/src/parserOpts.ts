import type { ParserOptions } from '@oriflame/conventional-changelog-types';
import { COMMIT_FORMAT_PREFIX } from '@oriflame/conventional-changelog-types';

const options: Partial<ParserOptions> = {
  headerCorrespondence: ['azure', 'type', 'scope', 'message'],
  // Keep in sync with checkCommitFormat
  headerPattern: new RegExp(`^(Merged? PR \\d+: )?${COMMIT_FORMAT_PREFIX.source} (.*)$`, 'u'),
  mergeCorrespondence: ['id', 'source'],
  mergePattern: /^Merge pull request #(\\d+) from (.*)/u,

  noteKeywords: ['BREAKING CHANGE', 'BREAKING CHANGES', 'Note'],
  revertCorrespondence: ['header'],
  revertPattern: /^Revert\s"(.*)"?/u,
};

export default options;
