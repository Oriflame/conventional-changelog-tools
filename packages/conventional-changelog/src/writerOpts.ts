/* eslint-disable @typescript-eslint/no-non-null-assertion -- needed */
/* eslint-disable no-console -- needed */
/* eslint-disable no-param-reassign -- needed */

import { createAzureClient, getWorkItemUrl } from '@oriflame/azure-helpers';
import {
  CommitGroupLabel,
  Context,
  GROUPS,
  Reference,
  WriterOptions,
} from '@oriflame/conventional-changelog-types';
import fs from 'fs';
import path from 'path';

import getTypeGroup from './getTypeGroup';

type GroupMap<T> = { [K in CommitGroupLabel]: T };

const groupEmojis = GROUPS.reduce(
  (data, group) => ({
    ...data,
    [group.label]: group.emoji,
  }),
  {},
) as GroupMap<string>;

const sortWeights: GroupMap<number> = {
  Release: 4,
  Breaking: 3,
  Updates: 2,
  Fixes: 1,
  Security: 0,
  Styles: -1,
  Types: -1,
  Docs: -2,
  Dependencies: -3,
  Misc: -3,
  Reverts: -4,
  Internals: -5,
};

const { SYSTEM_ACCESSTOKEN, ENDPOINT_URL_SYSTEMVSSCONNECTION } = process.env;

function* createWorkItemLink(workItemId: string) {
  createAzureClient({
    pat: SYSTEM_ACCESSTOKEN!,
    serverUrl: ENDPOINT_URL_SYSTEMVSSCONNECTION!,
  }).then((webApi) => {
    webApi.getWorkItemTrackingApi().then((workApi) => {
      workApi.getWorkItem(Number(workItemId)).then((workItem) => {
        if (workItemId) {
          yield getWorkItemUrl({ workItem, organizationUrl: ENDPOINT_URL_SYSTEMVSSCONNECTION! });
        }
      });
    });
  });

  return '';
}

function createLink(paths: string[], context: Context, reference: Partial<Reference> = {}): string {
  const owner = reference.owner ?? context.owner;
  const repository = reference.repository ?? context.repository;
  const url: string[] = [];

  if (repository) {
    if (context.host) {
      url.push(context.host);
    }

    if (owner) {
      url.push(owner);
    }

    url.push(repository);
  } else {
    url.push(context.repoUrl);
  }
  let base = url.join('/');

  // If deep linking to a sub-folder (monorepo project, etc),
  // extract the base URL if possible.
  [
    // github, gitlab
    'tree',
    'blob',
    // bitbucket
    'src',
  ].forEach((browsePart) => {
    if (base.includes(`/${browsePart}/`)) {
      [base] = base.split(`/${browsePart}/`);
    }
  });

  return [base, ...paths].join('/');
}

const options: Partial<WriterOptions> = {
  mainTemplate: fs.readFileSync(path.join(__dirname, '../templates/template.hbs'), 'utf-8'),
  commitPartial: fs.readFileSync(path.join(__dirname, '../templates/commit.hbs'), 'utf-8'),
  headerPartial: fs.readFileSync(path.join(__dirname, '../templates/header.hbs'), 'utf-8'),
  footerPartial: fs.readFileSync(path.join(__dirname, '../templates/footer.hbs'), 'utf-8'),

  // Commits
  groupBy: 'label',
  commitsSort: ['scope', 'message'],
  commitGroupsSort(groupA, groupB) {
    const aWeight = sortWeights[groupA.title] || 0;
    const bWeight = sortWeights[groupB.title] || 0;

    if (aWeight === 0 && bWeight === 0) {
      return groupA.title.localeCompare(groupB.title);
    }

    return bWeight - aWeight;
  },

  // Notes
  noteGroupsSort: 'title',

  // Add metadata
  transform(commit, context) {
    context.groupEmojis = groupEmojis;

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- false positive
    if (!commit.type) {
      return undefined;
    }

    // Use consistent values for snapshots
    if (process.env.NODE_ENV === 'test') {
      commit.hash = 'a1b2c3d';
      context.date = '2019-02-26';
    }

    // Override type for specific scenarios
    if (commit.revert) {
      commit.type = 'revert';
    } else if (commit.merge) {
      commit.type = 'misc';
    }

    // Define metadata based on type
    const group = getTypeGroup(commit.type);

    commit.label = group.label;

    if (group.bump === 'major') {
      context.isMajor = true;
    } else if (group.bump === 'minor') {
      context.isMinor = true;
    }

    if (context.commit.endsWith('s')) {
      // Workaround for azure devops
      commit.hashLink = createLink(['commit', commit.hash], context);
    } else {
      // Pre-generate links instead of doing it in handlebars
      commit.hashLink = createLink([context.commit, commit.hash], context);
    }

    // Use shorthand hashes
    if (typeof commit.hash === 'string') {
      commit.hash = commit.hash.slice(0, 7);
    }

    commit.references.forEach((reference) => {
      if (SYSTEM_ACCESSTOKEN) {
        reference.issueLink = createWorkItemLink(reference.issue).next().value;
      } else {
        reference.issueLink = createLink([context.issue, reference.issue], context, reference);
      }

      let source = `${reference.repository ?? ''}#${reference.issue}`;

      if (reference.owner) {
        source = `${reference.owner}/${source}`;
      }

      reference.source = source;
    });

    // Link users
    if (context.host) {
      commit.message = commit.message.replace(
        /\B@([a-z0-9](?:-?[a-z0-9/]){0,38})/gu,
        (match, username: string, index: number) => {
          if (
            username.includes('/') ||
            // Avoid when wrapped in backticks (inline code)
            commit.message.charAt(index - 1) === '`' ||
            commit.message.charAt(index + match.length + 1) === '`'
          ) {
            return match;
          }

          return `[@${username}](${context.host}/${username})`;
        },
      );
    }

    return commit;
  },
};

export default options;
