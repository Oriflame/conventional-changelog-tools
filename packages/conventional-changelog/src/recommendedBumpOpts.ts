import type { SemverLevel, Group } from '@oriflame/conventional-changelog-types';

import getTypeGroup from './getTypeGroup';
import parserOpts from './parserOpts';
import type { BumpOptions } from './types';

const options: BumpOptions = {
  parserOpts,

  whatBump(commits) {
    let level: SemverLevel = null;
    let breakings = 0;
    let features = 0;
    let fixes = 0;

    commits.forEach((commit) => {
      let group: Group;

      try {
        group = getTypeGroup(commit.type);
      } catch {
        return;
      }

      if (group.bump === 'major') {
        breakings += 1;
        level = 0;
      } else if (group.bump === 'minor') {
        features += 1;
        if (level === null || level === 2) {
          level = 1;
        }
      } else if (group.bump === 'patch') {
        fixes += 1;
        if (level === null) {
          level = 2;
        }
      }
    });

    return {
      level,
      reason: `There are ${breakings} breaking changes and ${features} new features, also ${fixes} fixes`,
    };
  },
};

export default options;
