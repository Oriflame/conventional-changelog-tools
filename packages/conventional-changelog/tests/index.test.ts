/* eslint-disable @typescript-eslint/no-unsafe-argument -- false positive */

/* eslint-disable @typescript-eslint/no-unsafe-call -- needed */
/* eslint-disable import/no-extraneous-dependencies -- needed */

import conventionalChangelogCore from 'conventional-changelog-core';
import conventionalRecommendedBump from 'conventional-recommended-bump';
import { writeFileSync } from 'fs';
import path from 'path';
import shell from 'shelljs';
import type Stream from 'stream';

import { config } from '../src';

function gitDummyCommit(msg: string[] | string, silent = true) {
  const args: string[] = ['--allow-empty', '--no-gpg-sign'];

  if (Array.isArray(msg)) {
    msg.forEach((m) => {
      args.push(`-m"${m}"`);
    });
  } else {
    args.push(`-m"${msg}"`);
  }

  shell.exec(`git commit ${args.join(' ')}`, { silent });
}

async function captureStreamOutput(stream: Stream.Readable) {
  return new Promise<string>((resolve, reject) => {
    let data = '';
    stream
      .on('error', (error: Error) => {
        reject(error);
      })
      .on('data', (chunk: string) => {
        data += String(chunk);
      })
      .on('end', () => {
        resolve(data.trim());
      });
  });
}

async function conventionalRecommendedBumpSteam(conventionalConfig: object) {
  return new Promise<Record<string, unknown>>((resolve, reject) => {
    conventionalRecommendedBump(
      {
        ...conventionalConfig,
      },
      (error: Error | null, result: Record<string, unknown>) => {
        console.log('recommended bump done');
        if (error !== null) {
          reject(error);
        }
        resolve(result);
      },
    );
  });
}

describe('conventional-changelog-ori', () => {
  const commonConfig = {
    config,
    pkg: {
      path: path.join(__dirname, 'package.json'),
    },
  };

  jest.setTimeout(60000);

  beforeEach(() => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any -- needed
    (shell.config as any).resetForTesting();
    shell.cd(__dirname);
    shell.mkdir('tmp');
    shell.cd('tmp');
    shell.mkdir('git-templates');
    shell.exec('git init --initial-branch master --template=./git-templates');
    shell.exec('git remote add origin https://github.com/user/repo.git');
    writeFileSync(
      'package.json',
      '{ "name": "@oriflame/conventional-changelog", "repository": { "type": "git", "url": "https://github.com/oriflame/conventional-changelog-tools.git" } }',
    );
    shell.exec('git add --all && git commit -m"First commit"');
  });

  afterEach(() => {
    shell.cd(__dirname);
    shell.rm('-rf', 'tmp');
  });

  it('supports all types at once', async () => {
    gitDummyCommit(['release: New major!', 'Note: New build system.']);
    gitDummyCommit(['break: Forms have changed', 'Note: They are easier now!']);
    gitDummyCommit(['new: amazing new module', 'Not backward compatible.']);
    gitDummyCommit('fix: updated i18n');
    gitDummyCommit(['update(modal, button): added accessibility', 'closes #1, #2']);
    gitDummyCommit('feature(core): settings refactor');
    gitDummyCommit('Random commit with no type');
    gitDummyCommit('docs: added getting started');
    gitDummyCommit('style(button): polished rounded corners');
    gitDummyCommit(['security(auth): improved logic', 'fixes #3']);
    gitDummyCommit('Revert PR #1');
    gitDummyCommit('ci(travis): fixed yaml config');
    gitDummyCommit('build(deps): updated dev tools');
    gitDummyCommit('test: setup testing framework');
    gitDummyCommit('internal(ts): updated types');
    gitDummyCommit('deps(babel,jest): Bumped to latest');
    gitDummyCommit(['patch(router): Fix params']);
    gitDummyCommit('types: Removed any');
    gitDummyCommit('Merged PR 21884: fix: Type of lastAddedCustomerOrderIds');
    gitDummyCommit('Merged PR 21884: new(Test): Hello darkness my old friend');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('works if there is no semver tag', async () => {
    gitDummyCommit(['build: first build setup', 'Note: New build system.']);
    gitDummyCommit(['ci(travis): add TravisCI pipeline', 'Continuously integrated.']);
    gitDummyCommit(['new: amazing new module', 'Not backward compatible.']);
    gitDummyCommit(['fix(compile): avoid a bug', 'The Change is huge.']);
    gitDummyCommit(['update(ngOptions): make it faster', 'closes #1, #2']);
    gitDummyCommit('revert(ngOptions): bad commit');
    gitDummyCommit('fix(*): oops');
    gitDummyCommit('type: Added unknown');
    gitDummyCommit('tests: Added before hooks');
    gitDummyCommit('Merged PR 21884: fix: Type of lastAddedCustomerOrderIds');
    gitDummyCommit('Merged PR 21884: new(Test): Hello darkness my old friend');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('works if there is a semver tag', async () => {
    shell.exec('git tag v1.0.0');
    gitDummyCommit('update: some more features');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
        outputUnreleased: true,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('works with unknown host', async () => {
    gitDummyCommit('docs: add manual');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
        pkg: {
          path: path.join(__dirname, 'package-unknown-repo.json'),
        },
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('uses h1 for major versions', async () => {
    gitDummyCommit('break: new shit');
    gitDummyCommit('release: new stuff');
    gitDummyCommit('fix: just a patch');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('uses h2 for minor versions', async () => {
    gitDummyCommit('new: new shit');
    gitDummyCommit('update: new stuff');
    gitDummyCommit('feature(modal): better modals');
    gitDummyCommit('fix: just a patch');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('uses h3 for patch versions', async () => {
    gitDummyCommit('docs: add a manual');
    gitDummyCommit('patch: just a patch');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it('replaces #[0-9]+ with issue URL', async () => {
    gitDummyCommit(['new(awesome): fix #88']);

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('replaces @username with GitHub user URL', async () => {
    gitDummyCommit(['feature(awesome): issue brought up by @bcoe! on Friday']);

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('doesnt replace @username if wrapped in backticks', async () => {
    gitDummyCommit(['deps: Updated \\`@types\\` packages.']);

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('handles multiple notes', async () => {
    gitDummyCommit(['release: Initial release', 'Note: Made a lot of changes']);
    gitDummyCommit(['fix(button): Made button changes', 'Note: Button is more buttony']);

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('links commits/issues to deep repositories correctly', async () => {
    gitDummyCommit(['update: supports sub-package links', ' closes #10']);

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
        pkg: {
          path: path.join(__dirname, 'package-monorepo.json'),
        },
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('supports non public GitHub repository locations', async () => {
    gitDummyCommit(['update(events): implementing #5 by @dlmr', ' closes #10']);
    gitDummyCommit('new: why this work?');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
        pkg: {
          path: path.join(__dirname, 'package-custom-repo.json'),
        },
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('only replaces with link to user if it is an username', async () => {
    gitDummyCommit(['fix: use npm@5 (@username)']);
    gitDummyCommit([
      'build(deps): bump @dummy/package from 7.1.2 to 8.0.0',
      'break: The Change is huge.',
    ]);

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('handles merge commits', async () => {
    gitDummyCommit(['fix: use yarn']);
    gitDummyCommit('Merge pull request #29 from owner/repo');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  it.skip('handles revert type', async () => {
    gitDummyCommit('revert(foo): undo this');
    gitDummyCommit('Revert this is the PR title');

    const data = await captureStreamOutput(
      conventionalChangelogCore({
        ...commonConfig,
      }),
    );

    expect(data).toMatchSnapshot();
  });

  describe('recommended bump', () => {
    test.skip.each(['break', 'breaking', 'release'])(
      'bumps major version for %s',
      async (major) => {
        gitDummyCommit(`${major}: new stuff`);
        gitDummyCommit(`${major}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);

        expect(result).toEqual({
          level: 0,
          reason: 'There are 2 breaking changes and 0 new features, also 0 fixes',
          releaseType: 'major',
        });
      },
    );

    test.skip.each(['break', 'breaking', 'release'])(
      'bumps major version for %s with azure devops prefix',
      async (major) => {
        gitDummyCommit(`Merged PR 21884: ${major}: new stuff`);
        gitDummyCommit(`Merged PR 21884: ${major}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);

        expect(result).toEqual({
          level: 0,
          reason: 'There are 2 breaking changes and 0 new features, also 0 fixes',
          releaseType: 'major',
        });
      },
    );

    test.skip.each(['new', 'update', 'feature'])('bumps minor version for %s', async (minor) => {
      gitDummyCommit(`${minor}: new stuff`);
      gitDummyCommit(`${minor}(todo): with scope`);
      const result = await conventionalRecommendedBumpSteam(commonConfig);

      expect(result).toEqual({
        level: 1,
        reason: 'There are 0 breaking changes and 2 new features, also 0 fixes',
        releaseType: 'minor',
      });
    });

    test.skip.each(['new', 'update', 'feature'])(
      'bumps minor version for %s with azure devops prefix',
      async (minor) => {
        gitDummyCommit(`Merged PR 21884: ${minor}: new stuff`);
        gitDummyCommit(`Merged PR 21884: ${minor}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);

        expect(result).toEqual({
          level: 1,
          reason: 'There are 0 breaking changes and 2 new features, also 0 fixes',
          releaseType: 'minor',
        });
      },
    );

    test.skip.each(['fix', 'deps', 'style', 'security', 'revert', 'misc', 'type', 'types'])(
      'bumps patch version for %s',
      async (patch) => {
        gitDummyCommit(`${patch}: new stuff`);
        gitDummyCommit(`${patch}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);

        expect(result).toEqual({
          level: 2,
          reason: 'There are 0 breaking changes and 0 new features, also 2 fixes',
          releaseType: 'patch',
        });
      },
    );

    test.skip.each(['fix', 'deps', 'style', 'security', 'revert', 'misc', 'type', 'types'])(
      'bumps patch version for %s with devops prefix',
      async (patch) => {
        gitDummyCommit(`Merged PR 21884: ${patch}: new stuff`);
        gitDummyCommit(`Merged PR 21884: ${patch}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);

        expect(result).toEqual({
          level: 2,
          reason: 'There are 0 breaking changes and 0 new features, also 2 fixes',
          releaseType: 'patch',
        });
      },
    );

    test.skip.each(['docs', 'ci', 'build', 'test', 'tests', 'internal'])(
      "doesn't bump version for %s",
      async (ignore) => {
        gitDummyCommit(`${ignore}: new stuff`);
        gitDummyCommit(`${ignore}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);

        expect(result).toEqual({
          level: null,
          reason: 'There are 0 breaking changes and 0 new features, also 0 fixes',
        });
      },
    );

    test.skip.each(['docs', 'ci', 'build', 'test', 'tests', 'internal'])(
      "doesn't bump version for %s with devops prefix",
      async (ignore) => {
        gitDummyCommit(`Merged PR 21884: ${ignore}: new stuff`);
        gitDummyCommit(`Merged PR 21884: ${ignore}(todo): with scope`);
        const result = await conventionalRecommendedBumpSteam(commonConfig);
        expect(result).toEqual({
          level: null,
          reason: 'There are 0 breaking changes and 0 new features, also 0 fixes',
        });
      },
    );

    it('does nothing when no type exist', async () => {
      gitDummyCommit('new stuff');
      gitDummyCommit('commit without a type');
      const result = await conventionalRecommendedBumpSteam(commonConfig);

      expect(result).toEqual({
        level: null,
        reason: 'There are 0 breaking changes and 0 new features, also 0 fixes',
      });
    });
  });
});
