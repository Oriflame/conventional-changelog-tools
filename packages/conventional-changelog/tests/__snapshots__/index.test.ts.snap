// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conventional-changelog-ori supports all types at once 1`] = `
"# 1.0.0 - 2019-02-26

New build system.

They are easier now!

#### 🎉 Release

- New major! ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 💥 Breaking

- Forms have changed ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🚀 Updates

- **[core]** settings refactor ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- **[modal, button]** added accessibility ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d)), closes  _#1_
  _#2_

- **[Test]** Hello darkness my old friend ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- amazing new module ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🐞 Fixes

- Type of lastAddedCustomerOrderIds ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- **[router]** Fix params ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- updated i18n ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🔑 Security

- **[auth]** improved logic ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d)), closes  _#3_


#### ⚙️ Types

- Removed any ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🎨 Styles

- **[button]** polished rounded corners ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 📘 Docs

- added getting started ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 📦 Dependencies

- **[babel,jest]** Bumped to latest ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🛠 Internals

- **[ts]** updated types ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- setup testing framework ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- **[deps]** updated dev tools ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- **[travis]** fixed yaml config ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))"
`;

exports[`conventional-changelog-ori works if there is no semver tag 1`] = `
"## 1.0.0 - 2019-02-26

New build system.

#### 🚀 Updates

- **[ngOptions]** make it faster ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d)), closes  _#1_
  _#2_

- **[Test]** Hello darkness my old friend ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- amazing new module ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🐞 Fixes

- Type of lastAddedCustomerOrderIds ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- **[compile]** avoid a bug ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### ⚙️ Types

- Added unknown ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### ↩️ Reverts

- **[ngOptions]** bad commit ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))

#### 🛠 Internals

- Added before hooks ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- **[travis]** add TravisCI pipeline ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))
- first build setup ([a1b2c3d](https://github.com/rajzik/conventional-changelog-beemo/commit/a1b2c3d))"
`;
